07_errors/n0npax/store_test.go

```diff
@@ -0,0 +1,189 @@
+package main
+
+import (
+	"testing"
+
+	tassert "github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+)
+
+var (
+	puppy0 = func() Puppy {
+		return Puppy{
+			Breed:  "Type A",
+			Colour: "Grey",
+			Value:  300,
+		}
+	}
+	puppy1 = func() Puppy {
+		return Puppy{
+			Breed:  "Type B",
+			Colour: "Brown",
+			Value:  400,
+		}
+	}
+	puppyNegativeValue = func() Puppy {
+		return Puppy{
+			Breed:  "Type C",
+			Colour: "Red",
+			Value:  -1,
+		}
+	}
+	puppyNegativeID = func() Puppy {
+		return Puppy{
+			Breed:  "Type D",
+			Colour: "Blue",
+			Value:  100,
+			ID:     -1,
+		}
+	}
+)
+
+type storerSuite struct {
+	suite.Suite
+	store Storer
+}
+
+func (s *storerSuite) SetupTest() {
+	switch s.store.(type) {
+	case *SyncStore:
+		s.store = NewSyncStore()
+	case *MemStore:
+		s.store = NewMemStore()
+	default:
+		panic("Unrecognised storer implementation")
+	}
+	puppy := puppy0()
+	_, err := s.store.CreatePuppy(&puppy)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func TestStorer(t *testing.T) {
+	suite.Run(t, &storerSuite{store: NewMemStore()})
+	suite.Run(t, &storerSuite{store: NewSyncStore()})
+}
+
+func (s *storerSuite) TestCreatePuppySuccessful() {
+	assert := tassert.New(s.T())
+	newPuppy0, newPuppy1 := puppy0(), puppy1()
+	id0, err := s.store.CreatePuppy(&newPuppy0)
+	assert.NoError(err, "Creating puppy should be ok")
+	id1, err := s.store.CreatePuppy(&newPuppy1)
+	assert.NoError(err, "Creating puppy should be ok")
+	assert.Equal(id0, id1-1, "2nd id should be 1st +1, got %v and %v", id0, id1)
+}
+
+func (s *storerSuite) TestCreatePuppyNegativeValue() {
+	assert := tassert.New(s.T())
+	newPuppy := puppyNegativeValue()
+	_, err := s.store.CreatePuppy(&newPuppy)
+	assert.Error(err, "negative ID should cause an error")
+}
+
+func (s *storerSuite) TestReadPuppySuccessful() {
+	for i := 0; i < 5; i++ {
+		assert := tassert.New(s.T())
+		newPuppy := puppy0()
+		id, err := s.store.CreatePuppy(&newPuppy)
+		assert.NoError(err, "Creating puppy should be ok")
+
+		readPuppy, err := s.store.ReadPuppy(id)
+		if assert.NoError(err, "Should be able to read puppy0 from store") {
+			assert.Equal(&newPuppy, readPuppy, "store should return identic puppy")
+		}
+		assert.Equal(id, readPuppy.ID)
+	}
+}
+
+func (s *storerSuite) TestCreateDeleteCreate() {
+	assert := tassert.New(s.T())
+	newPuppy0, newPuppy1 := puppy0(), puppy1()
+	_, err := s.store.CreatePuppy(&newPuppy0)
```

Comment:
When I am doing things like this, I like to try and put them in an array and iterate to avoid doing the duplicate logic.

```go
puppies := []Puppy{puppy0(), puppy1()}
for i, _ := range puppies {
  _, err := s.store.CreatePuppy(puppies[i])
  assert.NoError(err)
}
```

Reply:
Thx will fix in lab 10